Split out ACID FAQ into raw data considerations and financial data considerations (#1061)
tigerbeetle/tigerbeetle: docs/FAQ.md

Yes. Let's discuss each part:

### Atomicity

> As part of replication, each operation is durably stored in a majority

Reviewer: ```suggestion
As part of replication, each operation is durably stored in a quorum
```

Due to flexible quorums, a quorum is not necessarily a majority.

Author: Fixed, thanks!

> As part of replication, each operation is durably stored in at least a
> quorum of replicas' Write-Ahead Log (WAL) before the primary will mark
> the operation as successful. Only over time do WAL entries make their
> way through a state machine and into the LSM storage layer.

Reviewer 2: ```suggestion
As part of replication, each operation is durably stored in at least a
quorum of replicas' Write-Ahead Logs (WAL) before the primary will acknowledge
the operation as committed. WAL entries are executed through the state machine business logic and the resulting state changes are stored in TigerBeetle's LSM-Forest local storage engine.
```

The WAL is what allows TigerBeetle to achieve atomicity and durability

> since once the WAL is the source of truth. If TigerBeetle crashes, the

Reviewer: "since once the WAL is the source of truth"

Typo? I'm not sure what this is supposed to say.

Author: Oops yes, typo. Reworded this paragraph to be clearer (about what I meant anyway).

> is replayed on startup against the last checkpoint on disk.

Reviewer 2: ```suggestion
is replayed at startup from the last checkpoint on disk.
```

> Financial atomicity goes further than this though: events and

Reviewer 2: ```suggestion
However, financial atomicity goes further than this: events and
```

To use the same structure as the next section.

transfers can be linked when created so they all succeed or fail
together.

### Consistency

TigerBeetle guarantees strict serializability. And at the cluster

> level, stale reads are not possible since all operations (i.e. reads,
> not just writes) go through consensus.

Reviewer 2: ```suggestion
level, stale reads are not possible since all operations (not only writes, but also reads) go through the global consensus protocol.
```

However, financial consistency requires more than this. TigerBeetle

> exposes a double-entry accounting API where money is never created or
> destroyed, only transferred between accounts. And transfer history is

Reviewer 2: ```suggestion
exposes a double-entry accounting API to guarantee that money cannot be created or
destroyed, but only transferred from one account to another. And transfer history is
```

To follow Einstein's wording a little more closely with the law of conservation of energy:

> "Energy cannot be created or destroyed, it can only be changed from one form to another"

immutable. You can read more about our consistency guarantees
[here](./design/consistency.md).

### Isolation

> All events within a batch (across all clients) are
> serialized at the state machine level. There are no concurrent
> events in TigerBeetle, even across clients.

Reviewer 2: ```suggestion
All client requests (and all events within a client request batch) are executed with the highest level of isolation, serially through the state machine, one after another, before the next operation begins. Counterintuitively, the use of batching and serial execution means that TigerBeetle can also provide this level of isolation optimally, without the cost of locks for all the individual events within a batch.
```

We also serialize at the global consensus protocol layer beneath.

### Durability

> Up until 2018, traditional DBMS durability has focused on the crash
> consistency model, however, Fsyncgate and [Protocol Aware

Reviewer 2: ```suggestion
Up until 2018, traditional DBMS durability has focused on the Crash
Consistency Model, however, Fsyncgate and [Protocol Aware
```

Recovery](https://www.usenix.org/conference/fast18/presentation/alagappan)
have shown that this model can lead to real data loss for users in the
wild. TigerBeetle therefore adopts an explicit storage
fault model, which we then verify and test with incredible levels of

> corruption, something which few distributed systems
> were ever designed to handle. Our emphasis on protecting Durability is

Reviewer 2: ```suggestion
corruption, something which few distributed systems historically
were designed to handle. Our emphasis on protecting Durability is
```

what sets TigerBeetle apart, not only as a ledger but as a DBMS.

However, absolute durability is impossible, because all hardware can
ultimately fail. Data we write today might not be available
tomorrow. TigerBeetle embraces limited disk reliability and maximizes
data durability in spite of imperfect disks. We actively work against

> such entropy by taking advantage of cluster-wide storage. A bit of

Reviewer: Is "bit" being used colloquially here, or as a literal 1/8-of-a-byte? If the former I'd use a different word.

Author: This was @matklad 's phrase. I also am not completely sure if it was meant as a literal bit or rather "a chunk".

Reviewer 3: Colloquial sense, but I now see that this is ambiguous. Maybe just call it a "record"? 

> A record can only be lost if it gets corrupted on all replica, and even in that case the system would safely halt.

Author: Switched to record. Thanks!
